### PR TITLE
Codylandry/apply logging to bootstrap logs

### PR DESCRIPTION
**Pull request by codylandry**
## Problem 
We have noticed that logging configurations aren't considered for logs during the bootstrapping phase of the agent startup.  This results in a high number of logs generated, especially when using nodejs lambda functions.  

## Proposed Release Notes
* Updated logger to delay logging until configuration is parsed. The logger will now queue all log entries that occur before the agent can parse the configuration.

## Details
This PR adds functionality to the `Logger` class to enqueue and later flush a queue of logs.  The idea is that during the bootstrapping phase, we enqueue logs, holding them in an array.  Then, once the config has been loaded, including logging configs, we flush the queue in the original log order.

### CODE REVIEW

**Comment by codylandry** [comment]
FYI, I'm emailing opensource@newrelic.com with some questions about the CLA.  Will sign asap.

File: lib/util/logger.js [hunk 1]
```diff
@@ -196,6 +211,10 @@ Logger.prototype._read = function _read() {
  * already converted the objects to strings.
  * Returns a boolean representing the status of the write
  * (success/failure)
+ *
+ * @param level
+ * @param args
+ * @param extra
```
Comment (codylandry) [inline lib/util/logger.js:217]: Looks like my editor decided to add some annotations here.  I can remove them if needed.

**Review by bizob2828 — CHANGES_REQUESTED** [review]
H @codylandry, thanks for taking a stab at solving this.  Our logger as you can see has a few rough edges.  This circular dependency put us in an awkward situation.  I haven't dug much into this problem but after reviewing your PR and the code paths, I'm wondering if you could possibly make this more automagical?  So instead of adding first class queueing methods you just detect if the logger has been configured and if not throw them in the queue.  Then once the logger has been configured, flush the queue.

More specifically in this method https://github.com/newrelic/node-newrelic/blob/main/lib/util/logger.js#L76.  

```js
if(!this.options.configured) {
  this.logQueue.unshift({ level: _level, args: arguments })
}

**Comment by codylandry** [comment]
> H @codylandry, thanks for taking a stab at solving this. Our logger as you can see has a few rough edges. This circular dependency put us in an awkward situation. I haven't dug much into this problem but after reviewing your PR and the code paths, I'm wondering if you could possibly make this more automagical? So instead of adding first class queueing methods you just detect if the logger has been configured and if not throw them in the queue. Then once the logger has been configured, flush the queue.
> 
> More specifically in this method https://github.com/newrelic/node-newrelic/blob/main/lib/util/logger.js#L76.
> 
> ```js
> if(!this.options.configured) {
>   this.logQueue.unshift({ level: _level, args: arguments })
> }
> ```

For sure!  Thanks for taking a look.  I think that's perfectly reasonable.

**Comment by codylandry** [comment]
@bizob2828 I made some changes based on your comments along with a few more test cases.  Looking forward to getting your feedback.

**Comment by bizob2828** [comment]
> @bizob2828 I made some changes based on your comments along with a few more test cases. Looking forward to getting your feedback.

Thanks @codylandry, I plan on digging into this today.

File: test/unit/logger.test.js [hunk 2]
```diff
@@ -47,6 +48,50 @@ tap.test('Logger', function (t) {
...
+    logger.debug('debug')
+    logger.info('info')
+    logger.warn('warn')
+    logger.error('error')
+    logger.fatal('fatal')
+    t.ok(logger.logQueue.length === 6, 'should have 6 logs in the queue')
+    t.end()
+  })
+
+  t.test('should not equeue logs when disabled', function (t) {
```
Comment (bizob2828) [inline test/unit/logger.test.js:63]:
```suggestion
  t.test('should not enqueue logs when disabled', function (t) {
```

File: lib/logger.js [hunk 3]
```diff
@@ -11,8 +11,10 @@ const fs = require('./util/unwrapped-core').fs
 // create bootstrapping logger
 module.exports = new Logger({
```
Comment (bizob2828) [inline lib/logger.js:12]: now that we aren't redefining the module.exports you could just create the instance of logger 1 line above then on line 34 just reference `logger.configure`.  Either way this was a great idea and really streamlines this logger

File: lib/logger.js [hunk 4]
```diff
@@ -28,8 +30,8 @@ if (config) {
     enabled: config.logging.enabled
   }
 
-  // create the "real" logger
-  module.exports = new Logger(options)
+  // configure logger
+  module.exports.configure(options)
```
Comment (bizob2828) [inline lib/logger.js:34]: now that we aren't redefining the export you can remove the `config.setLogger` below.

Comment (bizob2828) [reply lib/logger.js:34]: the one use case we intend to support which does not appear to have a test is we started out as `stream: process.stdout`.  This did 1 good thing and obviously 1 bad thing which you're addressing.  But what needs to be handled is if when we get the config it failed to parse we need to set the stream to process.stdout. I think that's as simple as adding 2 items here https://github.com/newrelic/node-newrelic/blob/main/lib/config/index.js#L1715
```_configInstance = Object.assign(defaultConfig(), { agent_enabled: false, logging: { enabled: true, filepath: process.stdout }})```

File: test/unit/logger.test.js [hunk 5]
```diff
@@ -47,6 +48,50 @@ tap.test('Logger', function (t) {
     t.end()
   })
 
+  t.test('should equeue logs until configured', function (t) {
```
Comment (bizob2828) [inline test/unit/logger.test.js:51]: nice test additions 👏🏻

**Review by bizob2828 — CHANGES_REQUESTED** [review]
@codylandry thanks for addressing my feedback.  we're very close on getting this landed.

File: lib/logger.js (hunk 4 above)
Comment (codylandry) [reply lib/logger.js]: Makes total sense.  Thanks for catching that.

**Comment by codylandry** [comment]
@bizob2828 Changes made.  All great comments, thanks.  

Quick note.  I emailed the address mentioned here, but still haven't heard back.  I need to know if I need to use the company CLA and if so, initiate that process.
https://inmarb2b.slack.com/archives/G013TCRKZN1/p1646156382623099

**Comment by bizob2828** [comment]
> > @bizob2828 I made some changes based on your comments along with a few more test cases. Looking forward to getting your feedback.
> 
> Thanks @codylandry, I plan on digging into this today.

@codylandry this looks good. I want to manually test some scenarios but I reached out to our open source dept and they said you can sign the individual CLA

File: lib/config/index.js [hunk 6]
```diff
@@ -1712,7 +1713,14 @@ function getOrCreateInstance() {
 
       // Config construction has potential to throw due to invalid settings.
       // This allows the agent to return a stub api without crashing the process.
-      _configInstance = Object.assign(defaultConfig(), { agent_enabled: false })
+      _configInstance = Object.assign(defaultConfig(), {
+        agent_enabled: false,
+        logging: {
+          enabled: true,
+          filepath: process.stdout
```
Comment (bizob2828) [inline lib/config/index.js:1720]: ok i finally figured out a way to test this and I'm glad I did.  this should be `stdout` not process.stdout.  process.stdout is already a WriteStream so it'll fail when trying to create the writeStream in lib/logger with

```sh
node:internal/fs/utils:670
    throw new ERR_INVALID_ARG_TYPE(propName, ['string', 'Buffer', 'URL'], path);
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string or an instance of Buffer or URL. Received an instance of WriteStream
    at new WriteStream (node:internal/fs/streams:325:5)
    at Object.createWriteStream (node:fs:2889:10)
    at Object.<anonymous> (/Users/revans/code/node-newrelic/lib/logger.js:50:21)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/revans/code/node-newrelic/index.js:17:14) {
  code: 'ERR_INVALID_ARG_TYPE'
}
```

I also realized we lack tests for lib/logger but i'll take care of that in a different PR.

Comment (bizob2828) [reply lib/config/index.js:1720]: A way to manually test this is to add these 2 lines in a config

```
  high_security: true,
  security_policies_token: 'ffff',
```

I found that out by searching in the `test/unit/config` for t.throws and it found 

```js
tap.test('should throw with both high_security and security_policies_token defined', (t) => {
  t.throws(function testInitialize() {
    Config.initialize({
      high_security: true,
      security_policies_token: 'fffff'
    })
  })

  t.end()
})
```

Comment (codylandry) [reply lib/config/index.js]: I'm glad you tested it too!  Great catch.  I should've added a test case for that fallback config to begin with.  I made the change and added a quick test case that proves that the logger does get enabled and piping to stdout when the config is invalid.

**Review by bizob2828 — APPROVED** [review]

File: lib/config/index.js (hunk 6 above)
Comment (bizob2828) [reply lib/config/index.js]: wow, didn't realize we had those tests.  I'm still new-ish on the team.  thanks for adding that test 🙏🏻

**Comment by codylandry** [comment]
Thanks for the help @bizob2828.  Not sure if your team handles this repo as well, but thought I'd let you know I've submitted another logging related [PR in the lambda extension repo.](https://github.com/newrelic/newrelic-lambda-extension/pull/105).

**Comment by bizob2828** [comment]
> Thanks for the help @bizob2828. Not sure if your team handles the lambda extension repo as well, but thought I'd let you know I've submitted another logging related [PR there too](https://github.com/newrelic/newrelic-lambda-extension/pull/105).

No problem, thanks for opening the PR!  We do not maintain that but I can let the team that does know about it.

**Pull request merged**